backend/src/apiserver/storage/task_store.go

```diff
@@ -0,0 +1,210 @@
+			MLMDExecutionID:   mlmdExecutionID,
+			CreatedTimestamp:  createdTimestamp,
+			FinishedTimestamp: finishedTimestamp,
+			Fingerprint:       fingerprint,
+		}
+		tasks = append(tasks, task)
+	}
+	return tasks, nil
+}
+
+// Runs two SQL queries in a transaction to return a list of matching experiments, as well as their
+// total_size. The total_size does not reflect the page size.
+func (s *TaskStore) ListTasks(filterContext *common.FilterContext, opts *list.Options) ([]*model.Task, int, string, error) {
+	errorF := func(err error) ([]*model.Task, int, string, error) {
+		return nil, 0, "", util.NewInternalServerError(err, "Failed to list tasks: %v", err)
+	}
+
+	// SQL for getting the filtered and paginated rows
+	sqlBuilder := sq.Select(taskColumns...).From("tasks")
+	if filterContext.ReferenceKey != nil && filterContext.ReferenceKey.Type == common.Pipeline {
+		sqlBuilder = sqlBuilder.Where(sq.Eq{"PipelineName": filterContext.ReferenceKey.ID})
+	}
+	sqlBuilder = opts.AddFilterToSelect(sqlBuilder)
+
+	rowsSql, rowsArgs, err := opts.AddPaginationToSelect(sqlBuilder).ToSql()
+	if err != nil {
+		return errorF(err)
+	}
+
+	// SQL for getting total size. This matches the query to get all the rows above, in order
+	// to do the same filter, but counts instead of scanning the rows.
+	sqlBuilder = sq.Select("count(*)").From("tasks")
+	if filterContext.ReferenceKey != nil && filterContext.ReferenceKey.Type == common.Pipeline {
+		sqlBuilder = sqlBuilder.Where(sq.Eq{"PipelineName": filterContext.ReferenceKey.ID})
+	}
+	sizeSql, sizeArgs, err := opts.AddFilterToSelect(sqlBuilder).ToSql()
+	if err != nil {
+		return errorF(err)
+	}
+
+	// Use a transaction to make sure we're returning the total_size of the same rows queried
+	tx, err := s.db.Begin()
+	if err != nil {
+		glog.Errorf("Failed to start transaction to list jobs")
+		return errorF(err)
+	}
+
+	rows, err := tx.Query(rowsSql, rowsArgs...)
+	if err != nil {
+		tx.Rollback()
+		return errorF(err)
+	}
+	exps, err := s.scanRows(rows)
+	if err != nil {
+		tx.Rollback()
+		return errorF(err)
+	}
+	rows.Close()
+
+	sizeRow, err := tx.Query(sizeSql, sizeArgs...)
+	if err != nil {
+		tx.Rollback()
+		return errorF(err)
+	}
+	total_size, err := list.ScanRowToTotalSize(sizeRow)
+	if err != nil {
+		tx.Rollback()
+		return errorF(err)
+	}
+	sizeRow.Close()
+
+	err = tx.Commit()
+	if err != nil {
+		glog.Errorf("Failed to commit transaction to list experiments")
+		return errorF(err)
+	}
+
+	if len(exps) <= opts.PageSize {
+		return exps, total_size, "", nil
+	}
+
+	npt, err := opts.NextPageToken(exps[opts.PageSize])
+	return exps[:opts.PageSize], total_size, npt, err
+}
+
+func (s *TaskStore) GetTask(id string) (*model.Task, error) {
+	sql, args, err := sq.
+		Select(taskColumns...).
+		From("tasks").
+		Where(sq.Eq{"tasks.uuid": id}).
+		Limit(1).ToSql()
+	if err != nil {
+		return nil, util.NewInternalServerError(err, "Failed to create query to get pipeline: %v", err.Error())
```

Comment:
nit: to get task?
and other occurences below

Reply:
done